core/src/main/java/com/netflix/iceberg/util/BinPacking.java

```diff
@@ -107,7 +114,22 @@ public boolean hasNext() {
           bins.addLast(bin);
 
           if (bins.size() > lookback) {
-            return ImmutableList.copyOf(bins.removeFirst().items());
+            if (expensiveTaskFirst) {
```

Comment:
Would it make sense to place this logic into a separate method to keep `hasNext` short?

Reply:
With the current length of this block, I agree. But I think it could be much shorter using @rdsr's suggestions.

Reply:
+1 - Will do to keep implementation clean .